### ISSUE TITLE
Implement Native FlightSQL in IOx

### ISSUE THREAD

**Issue by alamb** [influxdata/influxdb_iox#6390]
Follow on on to the design hashed out in  https://github.com/influxdata/influxdb_iox/issues/6102

# Use Case

We would like to expand the tools available to our customers to execute queries using IOx. In particular, we would like IOx to support the [JDBC and ODBC drivers](https://arrow.apache.org/blog/2022/11/01/arrow-flight-sql-jdbc/), which are standard APIs for communicating with database servers. These drivers utilize the [FlightSQL](https://arrow.apache.org/blog/2022/02/16/introducing-arrow-flight-sql/) interface to interface with the database server.

# Design Doc
InfluxData only: https://docs.google.com/document/d/1MtasrcQ-iBGZ2WgiXK-Lk6DevygtuSewoihOWag0kGU/edit#

# Milestones
- [x] #6391
- [x] #6411
- [x] #6471
- [ ] ~#6789~
- [x] #7210

# Follow on work (moved out of scope from this initial project): 
- [ ] #6789
- [ ] #7265

# Post-Implementation Tasks
IOx supports both FlightSQL and its custom (1 round trip) Flight based SQL, and that will still be the case after this project
- [ ] ~Migrate Flight RPC clients (queryrouterd) in IDPE to FlightSQL~
- [x] Expose FlightSQL to customers
- [ ] ~Remove FlightRPC service from IOx~

**Comment by alamb** [comment]
Initial PR to start upstreaming flight client is here https://github.com/apache/arrow-rs/pull/3378

**Comment by alamb** [comment]
# Project Update

As of 2023-02-1 the `influxdb_iox` binary (as opposed to the IOx service exposed on InfluxDB cloud2) can run SQL via the FlightSQL protocol (for example, using the arrow flight JDBC drivers). The functionality is covered by [end to end tests](https://github.com/influxdata/influxdb_iox/blob/967450f0de6ba453f5c46477b795ac6f455bfa91/influxdb_iox/tests/end_to_end_cases/flightsql.rs#L183-L260)

The major known remaining  gaps in IOx's FlightSQL implementation are:
1. Support for parameterized queries (milestone 4)
2. Metadata API (milestone 5)

## Parameterized Queries

The term "parameterized query" means a query that has a placeholder like  `?` such as

```sql
SELECT * FROM cpu where time > ?
```

The values for these parameters are provided by the client in a separate API call.

There is a  circular dependency at the moment, where to implement support for parameterized queries I need  a FlightSQL client that supports them, to test the IOx implementation rather than implementing it "to spec". Unfortunately, neither the Rust FlightSQL client nor the JDBC driver supports parameterized statement yet (see the [jdbc ticket](https://github.com/apache/arrow/issues/33475) for more details)

To break this stalemate, I would like to find find either a FlightSQL client or server that has implemented parameterized queries.
* If I find a client I will implement and test IOx against this client.
* If I find a server, then I will make a rust client work with it, and then IOx work with the rust client.

It appears there is a reference implementation of a flight SQL server that supports parameterized statements here: https://github.com/apache/arrow/tree/master/go/arrow/flight/flightsql/example. This might be enough to implement the rust FlighSQL client against and I will propose doing so. 

I also asked on the mailing list https://lists.apache.org/thread/fr19z7xpsrxwkybr80zyx95hhjst3791

Work on some upstream pieces:
- [ ] Fix fallible streams in  arrow-flight client https://github.com/apache/arrow-rs/pull/3464
- [ ] Review UnionArray validation (which I think is used by parameterized statements): https://github.com/apache/arrow-rs/pull/1444
- [ ] Update the rust sql client in arrow to support parameterized statements
- [ ] Help @avantguardino with the JDBC driver implementation of parameterized statements: https://github.com/apache/arrow/pull/33961

## Metadata Queries

The JDBC driver appears to issue many of the metadata queries specified in FlightSQL such as `isReadOnly` (e.g. [here](https://github.com/influxdata/influxdb_iox/blob/967450f0de6ba453f5c46477b795ac6f455bfa91/influxdb_iox/tests/jdbc_client/Main.java#L122-L128)).

While the list of these types if queries is substantial ([source link](https://github.com/apache/arrow/blob/984248e6543fa7d05723c02a61a50447440ebd05/format/FlightSql.proto#L1116-L1479)) I think they are straightforward.

My plan here will be to implement one or two of them myself over the coming days, and then write tickets to solicit help with the implementation, as @jacob said there are people looking to help out.

**Comment by alamb** [comment]
Project Update:

After discussion with @gavincabbage
1.  We will postpone Milestone 4 (support for parameterized queries) until after Milestone 5 (metadata queries)
2. By the end of March 2023, I will implement one or two of the metadata queries for FlightSQL in the hopes that in April 2023 some of that work can be parallelized -- this is tracked by https://github.com/influxdata/idpe/issues/17197

**Comment by alamb** [comment]
Update: I am actively working on enough of the metadata queries (see https://github.com/influxdata/influxdb_iox/issues/7210) prior to parameterized queries

**Comment by alamb** [comment]
Update on this project: the current plan is that we will complete https://github.com/influxdata/influxdb_iox/issues/6789 and then claim success in this project. 

The major difference is that parameterized prepared statements will not be implemented (as they turned out not to be supported by the JDBC driver and not as important for actual tools)

I am tracking follow on work in  #6789
